### PR TITLE
[clang][Analyzer] Move checker 'alpha.unix.Errno' to 'unix.Errno'.

### PR DESCRIPTION
**Pull request by balazske**

### CODE REVIEW

File: clang/docs/analyzer/checkers.rst [hunk 1]
```diff
@@ -934,6 +934,76 @@ Check calls to various UNIX/Posix functions: ``open, pthread_once, calloc, mallo
...
+     // 'send' can be successful even if not all data was sent
+     if (errno == 1) { // An undefined value may be read from 'errno'
+       return 0;
+     }
+   }
+   return 1;
+ }
+
+The checker :ref:`unix-StdCLibraryFunctions` must be turned on to get the
+warnings from this checker. The supported functions are the same as by
```
Comment (DonatNagyE) [inline clang/docs/analyzer/checkers.rst:972]: If this checker doesn't do anything without unix.StdCLibraryFunctions, then this should be formalized as a dependency relationship (let this depend on unix.StdCLibraryFunctions). Are there any technical obstacles?

**Comment by balazske** [comment]
The checker was tested additionally on projects libwebm, bitcoin, contour and produced no results.

File: clang/docs/analyzer/checkers.rst (hunk 1 above)
Comment (DonatNagyE) [reply clang/docs/analyzer/checkers.rst:972]: Thanks for the clarification. I think it'd be important to improve this situation (it's very user-hostile to hide an "if you want to use this, you must manually enable another checker" remark in the docs), but this is independent of the currently reviewed commit.

**Comment by DonatNagyE** [comment]
This [report from vim](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=vim_v8.2.1920_errno_alpha_unix&newcheck=vim_v8.2.1920_errno_unix&is-unique=on&diff-type=New&report-id=3214372&report-hash=9582050be3f26e8ebf9dcf4c5b0f82b3&report-filepath=%2amemline.c) seems to be a false positive:

image

Readlink returns either -1 (failure, errno is set) or the number of characters read from the symlink entry (i.e. the filename the symlink points to). I don't think that it's possible to create a "symlink to emptystring", so the the checkers should recognize that if the return value is `<= 0`, then it's a failure.

**Comment by DonatNagyE** [comment]
I also think that the [reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3253666&report-hash=b4e0b723164236269fe6078ad32a0456&report-filepath=%2apg_basebackup.c) [from](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3254052&report-hash=619eb5d998324adb5e02d9e3302bb4d5&report-filepath=%2acopy.c) [postgres](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3256373&report-hash=32f8e213c6fb419277ec76c40bfa3956&report-filepath=%2afe-connect.c) that you mentioned are too confusing. They follow this pattern:

image

The function `fwrite()` is declared as
```c
size_t fwrite(const void *restrict ptr, size_t size, size_t nitems,
           FILE *restrict stream);
```
and it usually indicates success by returning the value of `nitems` -- but it has a corner case that if `size` is 0, then it does nothing and returns 0. This corner case is considered to be a successful call, and the bug reports are generated when the analyzer follows this case and notices that `errno` contains an undefined value.

I think that in these reports it would be important to replace the generic "`after successful call to 'fwrite'`" with a concrete message which highlight that the analyzer is following this obscure corner case (which is sometimes missing from the documentation, for example on my system `man 3 fwrite` from release 5.05 of the Linux man-pages project does not mention it).

**Review by NagyDonat — COMMENTED** [review]
I'd say that the `readlink` modeling tweak is not a blocking issue; but the bug reports coming from "size is zero" corner case should be clarified (in a separate commit) before merging this.

As these are both easy to fix, I hope that we can move this checker out of alpha soon.

**Comment by balazske** [comment]
PRs #71373 and #71392 are created to improve the indicated problems.

**Comment by balazske** [comment]
> I also think that the [reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3253666&report-hash=b4e0b723164236269fe6078ad32a0456&report-filepath=%2apg_basebackup.c) [from](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3254052&report-hash=619eb5d998324adb5e02d9e3302bb4d5&report-filepath=%2acopy.c) [postgres](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3256373&report-hash=32f8e213c6fb419277ec76c40bfa3956&report-filepath=%2afe-connect.c) that you mentioned are too confusing.
> 

After the change in #71392 the report looks like [this](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_test_71392_base&newcheck=postgres_REL_13_0_test_71392&is-unique=on&diff-type=Unresolved&checker-name=alpha.unix.Errno&report-hash=0b2d4e61a48cd556bb982b39969b81d2&report-id=3386353&report-filepath=%2acopy.c). This looks somewhat better, probably still false positive because the "len" can not be 0, but the checker does not have more information.

**Review by NagyDonat — APPROVED** [review]
> > I also think that the [reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3253666&report-hash=b4e0b723164236269fe6078ad32a0456&report-filepath=%2apg_basebackup.c) [from](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3254052&report-hash=619eb5d998324adb5e02d9e3302bb4d5&report-filepath=%2acopy.c) [postgres](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_errno_alpha_unix&newcheck=postgres_REL_13_0_errno_unix&is-unique=on&diff-type=New&report-id=3256373&report-hash=32f8e213c6fb419277ec76c40bfa3956&report-filepath=%2afe-connect.c) that you mentioned are too confusing.
> 
> After the change in #71392 the report looks like [this](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=postgres_REL_13_0_test_71392_base&newcheck=postgres_REL_13_0_test_71392&is-unique=on&diff-type=Unresolved&checker-name=alpha.unix.Errno&report-hash=0b2d4e61a48cd556bb982b39969b81d2&report-id=3386353&report-filepath=%2acopy.c). This looks somewhat better, probably still false positive because the "len" can not be 0, but the checker does not have more information.

Thanks for the update!

I think these improvements are sufficient to move this checker out of alpha.

The new message which clarifies that the report is produced in the case _"Assuming that argument 'size' to 'fwrite' is 0; 'errno' becomes undefined after the call"_ is completely sufficient to orient the user; and although the report is (arguably) a false positive, the user can quickly understand and discard it.

(This false positive is not too common and it's caused by the general limitations of the CSA engine, so it should not keep this checker in alpha state.)

@steakhal or anybody else interested:
Is there any objection to merging #71392 (the commit that clarifies the note tags) and this commit?

**Pull request merged**